[SPARK-8647][MLlib] Potential issue with constant hashCode

aloknsingh opened the pull request:
I added the code, 
  // see [SPARK-8647], this achieves the needed constant hash code without constant no.
  override def hashCode(): Int = this.getClass.getName.hashCode()

does getting the constant hash code as per jira

Commit:
[SPARK-8647][MLlib] Potential issue with constant hashCode

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
At best, this exchanges one number for another, so does that help anything? This changes semantics a bit if there are any subclasses. I don't think this is something to change without a clear motivation. At best a comment is warranted.

MechCoder commented:
Could you please explain what the issue is? Every instance of `MatrixUDT` should have the same hash code and it should not be a problem whatever this is.

mengxr commented:
I think @aloknsingh 's points are valid. Both `1994` and `7919` are magic numbers, which we should avoid in the code.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -193,7 +193,8 @@ private[spark] class MatrixUDT extends UserDefinedType[Matrix] {
 
-  override def hashCode(): Int = 1994
+  // see [SPARK-8647], this achieves the needed constant hash code without constant no.
+  override def hashCode(): Int = this.getClass.getName.hashCode()
```
Would `this.getClass.hashCode` be sufficient?

aloknsingh replied:
@mengxr agree I will change as per comment

aloknsingh replied:
@srowen @MechCoder Please see the jira https://issues.apache.org/jira/browse/SPARK-8647  for more details.

mengxr commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
[SPARK-8647][MLlib] to avoid the class derivation issues,
change the constant hashCode to
override def hashCode(): Int = classOf[MatrixUDT].getName.hashCode()

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36335 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36335/consoleFull) for   PR 7146 at commit [`e58bccf`](https://github.com/apache/spark/commit/e58bccfad20e826b3fcaa2dc9b637069e7abfa94).

SparkQA commented:
[Test build #36335 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36335/console) for   PR 7146 at commit [`e58bccf`](https://github.com/apache/spark/commit/e58bccfad20e826b3fcaa2dc9b637069e7abfa94).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

srowen commented:
LGTM in the sense that it just changes one number for another.

mengxr commented:
Merged into master. Thanks!
